simple_certmanager/csr_generation.py

```diff
@@ -0,0 +1,70 @@
+from cryptography import x509
+from cryptography.hazmat.backends import default_backend
+from cryptography.hazmat.primitives import hashes, serialization
+from cryptography.hazmat.primitives.asymmetric import rsa
+from cryptography.x509 import NameOID
+
+# These methods make use of the SigningRequest model to generate a private key and CSR
+# It is not imported here to avoid circular imports
+
+
+def generate_csr_and_private_key(signing_request):
+    generate_private_key(signing_request)
+
+    if not signing_request.csr:
+        # Generate CSR if not present
+        csr_builder = x509.CertificateSigningRequestBuilder()
+        csr_builder = generate_csr(signing_request, csr_builder)
+
+        # Load private key for signing CSR
+        private_key = serialization.load_pem_private_key(
+            signing_request.private_key.encode(),
+            password=None,
+            backend=default_backend(),
+        )
+
+        # Sign CSR with private key
+        csr = csr_builder.sign(private_key, hashes.SHA256(), default_backend())
+        # Only store the CSR bytes in the text field
+        signing_request.csr = csr.public_bytes(serialization.Encoding.PEM).decode()
+
+
+def generate_csr(signing_request, csr_builder):
+    csr_builder = csr_builder.subject_name(
+        x509.Name(
+            [
+                x509.NameAttribute(NameOID.COUNTRY_NAME, signing_request.country_name),
+                x509.NameAttribute(
+                    NameOID.STATE_OR_PROVINCE_NAME,
+                    signing_request.state_or_province_name,
+                ),
+                x509.NameAttribute(
+                    NameOID.LOCALITY_NAME, signing_request.organization_name
+                ),
+                x509.NameAttribute(
+                    NameOID.ORGANIZATION_NAME, signing_request.organization_name
+                ),
+                x509.NameAttribute(NameOID.COMMON_NAME, signing_request.common_name),
+                x509.NameAttribute(
+                    NameOID.EMAIL_ADDRESS, signing_request.email_address
+                ),
+            ]
+        )
+    )
+
+    return csr_builder
+
+
+def generate_private_key(signing_request):
+    if not signing_request.private_key:
+        # Generate private key
+        private_key = rsa.generate_private_key(
+            public_exponent=65537, key_size=4096, backend=default_backend()
+        )
+        private_key_file_bytes = private_key.private_bytes(
+            encoding=serialization.Encoding.PEM,
+            format=serialization.PrivateFormat.PKCS8,
+            encryption_algorithm=serialization.NoEncryption(),
+        )
+        # Only store the private key bytes in the text field
+        signing_request.private_key = private_key_file_bytes.decode()
```

Comment:
```suggestion
from typing import NewType

from cryptography import x509
from cryptography.hazmat.backends import default_backend
from cryptography.hazmat.primitives import hashes, serialization
from cryptography.hazmat.primitives.asymmetric import rsa
from cryptography.x509 import NameOID

PrivateKey = NewType("PrivateKey", str)
CSR = NewType("CSR", str)

# 4096 as "sane default". 2048 is considered too small these days, and the Ansible
# plays generate keys with this key size too.
KEY_SIZE_BITS = 4096


def generate_private_key() -> PrivateKey:
    """
    Generate an RSA private key and return the PEM-encoded key data.
    """
    private_key = rsa.generate_private_key(
        public_exponent=65537,
        key_size=KEY_SIZE_BITS,
        backend=default_backend(),
    )
    private_key_file_bytes = private_key.private_bytes(
        encoding=serialization.Encoding.PEM,
        format=serialization.PrivateFormat.PKCS8,
        encryption_algorithm=serialization.NoEncryption(),
    )
    # because it is PEM format, which is base64 encoding, we can safely decode the
    # bytes as ASCII
    return PrivateKey(private_key_file_bytes.decode("ascii"))


def generate_csr(
    key_pem: str,
    *,
    common_name: str,
    country: str = "",
    state_or_province: str = "",
    locality: str = "",
    organization_name: str = "",
    email: str = "",
) -> CSR:
    """
    Generate a CSR for the given private key and return the PEM-encoded CSR data.

    :arg key_pem: The private key to sign the CSR with, provided as PEM-encoded string.
    :arg common_name: The Subject common name, required.
    :arg country: The Subject country, as two-letter country code.
    :arg state_or_province: The Subject state or province name.
    :arg locality: The Subject locality (city) name.
    :arg organization_name: The Subject organization name.
    :arg email: The Subject email address.
    """
    assert common_name, "Common name may not be empty"

    # Load the private key to sign the CSR with.
    private_key = serialization.load_pem_private_key(
        key_pem.encode("ascii"),
        password=None,
        backend=default_backend(),
    )
    # type narrowing - we know we generate RSA private keys, see
    # :func:`generate_private_key`
    assert isinstance(private_key, rsa.RSAPrivateKey)

    # Only include the name attributes in the CSR that have non-empty values
    subject_name_attributes = [
        x509.NameAttribute(attr, value)
        for attr, value in [
            (NameOID.COMMON_NAME, common_name),
            (NameOID.ORGANIZATION_NAME, organization_name),
            (NameOID.LOCALITY_NAME, locality),
            (NameOID.STATE_OR_PROVINCE_NAME, state_or_province),
            (NameOID.COUNTRY_NAME, country),
            (NameOID.EMAIL_ADDRESS, email),
        ]
        if value
    ]
    csr_builder = x509.CertificateSigningRequestBuilder().subject_name(
        x509.Name(subject_name_attributes)
    )

    # Sign the CSR with our private key
    csr = csr_builder.sign(
        private_key, algorithm=hashes.SHA256(), backend=default_backend()
    )

    csr_pem = csr.public_bytes(serialization.Encoding.PEM).decode("ascii")
    return CSR(csr_pem)


def generate_private_key_with_csr(
    *,
    common_name: str,
    country: str = "",
    state_or_province: str = "",
    locality: str = "",
    organization_name: str = "",
    email: str = "",
) -> tuple[PrivateKey, CSR]:
    """
    Generate and return a pair of private key + CSR.

    :arg common_name: The Subject common name, required.
    :arg country: The Subject country, as two-letter country code.
    :arg state_or_province: The Subject state or province name.
    :arg locality: The Subject locality (city) name.
    :arg organization_name: The Subject organization name.
    :arg email: The Subject email address.
    """
    if not common_name:
        raise ValueError("You must provide a non-empty common_name.")

    private_key = generate_private_key()
    csr = generate_csr(
        key_pem=private_key,
        common_name=common_name,
        country=country,
        state_or_province=state_or_province,
        locality=locality,
        organization_name=organization_name,
        email=email,
    )
    return private_key, csr
```

This is roughly equivalent, the changes I make put emphasis on:

* no interaction with django at all
* type annotations
* simple functions with simple primitive arguments
* explicit keyword arguments instead of ambiguous positional arguments
* input validation (common name may not be empty), handling optional values by not including them
* added comments to document some design decisions

Of course, the model layer will need some adaption to interface with these functions, but those are the things that actually belong in the model layer :) See the matching code suggestions there.